jfix-stdlib-ratelimiter/src/main/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcher.kt

```diff
@@ -0,0 +1,282 @@
+package ru.fix.stdlib.ratelimiter
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.sync.Semaphore
+import org.slf4j.LoggerFactory
+import ru.fix.aggregating.profiler.PrefixedProfiler
+import ru.fix.aggregating.profiler.ProfiledCall
+import ru.fix.aggregating.profiler.Profiler
+import ru.fix.dynamic.property.api.DynamicProperty
+import ru.fix.dynamic.property.api.PropertySubscription
+import java.lang.invoke.MethodHandles
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.ForkJoinPool
+import java.util.concurrent.RejectedExecutionException
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Supplier
+import kotlin.coroutines.EmptyCoroutineContext
+
+/**
+ * Manages tasks execution with given rate and window.
+ * Rate specify how many requests per second will be dispatched.
+ * Window size specify how many async operations with uncompleted result allowed.
+ * When Window or Rate restriction is reached, dispatcher will stop to process requests and enqueue them in umbound queue.
+ * Dispatcher executes all operations in single dedicated thread.
+ *
+ * @param name           name of dispatcher - will be used in metrics and worker's thread name
+ * @param rateLimiter    rate limiter, which provides rate of operation
+ * @param windowSize     ho many async operations with uncompleted result are allowed in dispatcher
+ * @param closingTimeout max amount of time (in milliseconds) to wait for pending operations during shutdown.
+ *                       If parameter equals 0 then dispatcher will not wait pending operations during closing process.
+ *                       Any negative number will be interpreted as 0.
+ *
+ */
+class SuspendableRateLimitedDispatcher(
+        name: String,
+        rateLimiter: RateLimiter,
+        profiler: Profiler,
+        windowSize: DynamicProperty<Int>,
+        closingTimeout: DynamicProperty<Long>,
+        coroutineScope: CoroutineScope = DispatcherCommonPoolScope
+) : AutoCloseable {
+
+    companion object {
+        private const val QUEUE_SIZE_INDICATOR = "queue_size"
+        private const val ACTIVE_ASYNC_OPERATIONS = "active_async_operations"
+        private const val WINDOW_ACQUIRE_DELAY_MS = 500L
+    }
+
+    private val logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())
+
+    private val state = AtomicReference<State>()
+
+    private val rateLimiter: RateLimiter
+    private val deferreds = CopyOnWriteArrayList<Deferred<Any?>>()
```

Comment:
This list is frequently updated, thus CopyOnWriteArrayList is a very bad choice in terms of performance.

Actually, we do not need to store all executing coroutines to await/cancel them during close() method.

We need to use SupervisorJob() as a part of CoroutineContext, in which tasks must be executed. It also means, that we need to provide CoroutineContext to SuspendableRateLimitedDispatcher instead of CoroutineScope. 

Look at WaitingScope we have written once on our project.